subiquity/server/controllers/tests/test_filesystem.py

```diff
@@ -2343,3 +2345,68 @@ def test_is_reset_partition_only_utility(self, config, expected):
         self.app.autoinstall_config = config
 
         self.assertEqual(self.fsc.is_reset_partition_only(), expected)
+
+
+class TestGuidedChoiceValidation(IsolatedAsyncioTestCase):
+    def test_pin_and_pass(self):
+        reformat = GuidedStorageTargetReformat
+        tpmfde = GuidedCapability.CORE_BOOT_ENCRYPTED
+        choice = GuidedChoiceV2(
+            target=reformat, capability=tpmfde, pin="01234", password="asdf"
+        )
+        with self.assertRaises(StorageRecoverableError):
+            choice.validate()
+
+    @parameterized.expand(
+        (
+            (GuidedCapability.MANUAL, False, False),
+            (GuidedCapability.LVM_LUKS, False, True),
+            (GuidedCapability.CORE_BOOT_ENCRYPTED, True, True),
+        )
+    )
+    def test_capability_pin_pass_validation(self, capability, pin_ok, pass_ok):
+        def maybe_assert_raises(ok: bool):
+            if ok:
+                return contextlib.nullcontext()
+            else:
+                return self.assertRaises(StorageRecoverableError)
```

Comment:
praise: I really like this construction. I've written enough tests like:
```python
if ok:
    # do thing
else:
    with self.assertRaises(...):
        # do bad thing
```
and it makes me want to go fix them now.